cluster-api/pkg/controller/machineset/controller.go

```diff
@@ -0,0 +1,218 @@
+	// rate limited requeues on errors
+	queue workqueue.RateLimitingInterface
+}
+
+// New returns an instance of the MachineSet controller
+func New(machineClient machineclientset.Interface, machineSetsInformer cache.SharedIndexInformer, machineSetsLister machinelistersv1alpha1.MachineSetLister, machineLister machinelistersv1alpha1.MachineLister) *Controller {
+	controller := &Controller{
+		queue:               workqueue.NewNamedRateLimitingQueue(workqueue.NewItemFastSlowRateLimiter(2*time.Second, 10*time.Second, 5), "MachineSets"),
+		machineClient:       machineClient,
+		machineSetsInformer: machineSetsInformer,
+		machineSetsLister:   machineSetsLister,
+		machineLister:       machineLister,
+	}
+
+	controller.machineSetsInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err == nil {
+				controller.queue.AddRateLimited(key)
+			}
+		},
+		UpdateFunc: func(old interface{}, new interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(new)
+			if err == nil {
+				controller.queue.AddRateLimited(key)
+			}
+		},
+	})
+
+	return controller
+}
+
+// Run starts the control loop. This method is blocking.
+func (c *Controller) Run(threadiness int, stopCh <-chan struct{}) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	glog.Info("starting machineset controller")
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	glog.Info("shutting down machineset controller")
+}
+
+func (c *Controller) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false
+// when it's time to quit.
+func (c *Controller) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.syncHandler(key.(string))
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	runtime.HandleError(fmt.Errorf("%v failed with: %v", key, err))
+	c.queue.AddRateLimited(key)
+
+	return true
+}
+
+// syncHandler holds the controller's business logic.
+// it makes sure that the current state is equal to the desired state.
+// note that the current state of the cluster is calculated based on the number of machines
+// that are owned by the given machineSet (key).
+func (c *Controller) syncHandler(key string) error {
+	machineSet, err := c.machineSetsLister.Get(key)
+	if err != nil {
+		return err
+	}
+
+	filteredMachines, err := c.getMachines(machineSet)
+	if err != nil {
+		return err
+	}
+
+	return c.syncReplicas(machineSet, filteredMachines)
+}
+
+// syncReplicas essentially scales machine resources up and down.
+func (c *Controller) syncReplicas(machineSet *v1alpha1.MachineSet, machines []*v1alpha1.Machine) error {
+	currentMachineCount := int32(len(machines))
+	desiredReplicas := int32(0)
+	if machineSet.Spec.Replicas != nil {
```

Comment:
The default value for this field should be handled by the API server. It should never be `nil`

Reply:
the last time I checked the API server didn't set the default value. I am not sure if this is a bug or we should actually employ a mutating webhook for that. Note that we are using `CRDs`

Reply:
We don't have any defaulters and validation yet. For CRDs this will be `nil` unless we add mutating webhook

Reply:
Once it's moved to api aggregation world. We can have defaulting implemented easily.